scripts/o3de/o3de/compatibility.py

```diff
@@ -0,0 +1,295 @@
+    """
+    Returns True if the gem is compatible with the gems registered with the engine and optional project.
+    :param gem_json_data: gem json data dictionary
+    :param project_path: path to the project (optional)
+    :param gem_paths: paths to all gems to use for dependency checks (optional)
+    :param check: if True always return True and ouput log info about the result
+    """
+    # try to avoid gem compatibility checks which incur the cost of 
+    # opening many gem.json files to get version information
+    gem_dependencies = gem_json_data.get('dependencies','')
+    if not gem_dependencies:
+        return True
+
+    incompatible_gem_version_specifiers = None
+
+    if not gem_paths:
+        gem_paths = manifest.get_engine_gems()
+        gem_paths.extend(manifest.get_project_gems(project_path))
+        # Convert each path to pathlib.Path object and filter out duplicates using dict.fromkeys
+        gem_paths = list(dict.fromkeys(map(lambda gem_path_string: pathlib.Path(gem_path_string), gem_paths)))
+
+    incompatible_gem_version_specifiers = get_incompatible_gem_version_specifiers(
+        project_path, gem_dependencies, gem_paths)
+    if incompatible_gem_version_specifiers:
+        if check:
+            logger.info(f'{gem_json_data["gem_name"]} is not known to be compatible with the '
+            ' following gems used by this project and requires the --force parameter to enable.'
+            f'{incompatible_gem_version_specifiers}')
+        else:
+            logger.error(f'{gem_json_data["gem_name"]} is not known to be compatible with the '
+            ' following gems used by this project and requires the --force parameter to enable.'
+            f'{incompatible_gem_version_specifiers}')
+            return False
+    return True
+
+
+def gem_project_compatible(gem_json_data:dict, project_path:pathlib.Path, gem_paths:list = [], check:bool = False) -> bool:
+    """
+    Returns True if the gem is compatible with the indicated project.
+    :param gem_json_data: gem json data dictionary
+    :param project_path: path to the project
+    :param gem_paths: paths to all gems to use for dependency checks (optional)
+    :param check: if True always return True and ouput log info about the result
+    """
+    project_json_data = manifest.get_project_json_data(project_path=project_path)
+    if not project_json_data:
+        logger.error(f'Failed to load project.json data from {project_path} needed for checking compatibility')
+        return False
+
+    engine_name = project_json_data.get('engine','')
+    if engine_name:
+        engine_json_data = manifest.get_engine_json_data(engine_name=engine_name)
+    else:
+        engine_json_data = manifest.get_engine_json_data(engine_path=manifest.get_this_engine_path())
+    if not engine_json_data:
+        logger.error(f'Failed to load engine.json data based on the engine field in project.json or detect the engine from the current folder')
+        return False
```

Comment:
Actually the [get_project_engine_path](https://github.com/o3de/o3de/blob/development/scripts/o3de/o3de/manifest.py#L309-L311) should be used here to get the path to the engine the project is actually using and then manifest.get_engine_json_data should be called with that path.